### ISSUE TITLE
azurerm azurerm_network_interface => diffs didn't match during apply

### ISSUE THREAD

**Issue by pgroene** [hashicorp/terraform#7553]
### Terraform Version
Terraform v0.6.16

### Affected Resource(s)
- azurerm_network_interface

### Terraform Configuration Files
```hcl
variable "count" { default = "1" }
variable "location" { default = "West Europe" }
variable "name" { default = "trn12323" }

# Configure the Azure Resource Manager Provider
provider "azurerm" {
  subscription_id = "${var.subscription_id}"
  client_id       = "${var.client_id}"
  client_secret   = "${var.client_secret}"
  tenant_id       = "${var.tenant_id}"
}

resource "azurerm_resource_group" "trainingvms" {
    name = "${var.name}resourcegroup"
    location = "${var.location}"
}

resource "azurerm_virtual_network" "vnet" {
    name = "${var.name}vnet"
    address_space = ["10.0.0.0/16"]
    location = "${var.location}"
    resource_group_name = "${azurerm_resource_group.trainingvms.name}"
}

resource "azurerm_subnet" "subnet" {
    name = "${var.name}subnet"
    resource_group_name = "${azurerm_resource_group.trainingvms.name}"
    virtual_network_name = "${azurerm_virtual_network.vnet.name}"
    address_prefix = "10.0.2.0/24"
}


resource "azurerm_public_ip" "publicip" {
    count = "${var.count}"
    name = "${var.name}${count.index}"
    location = "${var.location}"
    resource_group_name = "${azurerm_resource_group.trainingvms.name}"
    public_ip_address_allocation = "static"
}


resource "azurerm_network_interface" "nic" {
    count = "${var.count}"
    name = "${var.name}nic${count.index}"
    location = "${var.location}"
    resource_group_name = "${azurerm_resource_group.trainingvms.name}"

    ip_configuration {
        name = "testconfiguration1"
        subnet_id = "${azurerm_subnet.subnet.id}"
        private_ip_address_allocation = "dynamic"
        public_ip_address_id = "${element(azurerm_public_ip.publicip.*.id, count.index)}"
    }
}
```

### Debug Output
Error applying plan:

1 error(s) occurred:

* azurerm_network_interface.nic.0: diffs didn't match during apply. This is a bug with Terraform and should be reported as a GitHub Issue.

Please include the following information in your report:

    Terraform Version: 0.6.16
    Resource ID: azurerm_network_interface.nic.0
    Mismatch reason: attribute mismatch: ip_configuration.~463970659.name
    Diff One (usually from plan): *terraform.InstanceDiff{Attributes:map[string]*terraform.ResourceAttrDiff{"ip_configuration.~463970659.private_ip_address_allocation":*terraform.ResourceAttrDiff{Old:"", New:"dynamic", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Type:0x0}, "ip_configuration.~463970659.load_balancer_backend_address_pools_ids.#":*terraform.ResourceAttrDiff{Old:"", New:"", NewComputed:true, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Type:0x0}, "ip_configuration.463970659.private_ip_address_allocation":*terraform.ResourceAttrDiff{Old:"dynamic", New:"", NewComputed:false, NewRemoved:true, NewExtra:interface {}(nil), RequiresNew:false, Type:0x0}, "ip_configuration.~463970659.subnet_id":*terraform.ResourceAttrDiff{Old:"", New:"/subscriptions/9bdde4e2-946e-4961-b111-fadcea94577d/resourceGroups/trn12323resourcegroup/providers/Microsoft.Network/virtualNetworks/trn12323vnet/subnets/trn12323subnet", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Type:0x0}, "ip_configuration.463970659.subnet_id":*terraform.ResourceAttrDiff{Old:"/subscriptions/9bdde4e2-946e-4961-b111-fadcea94577d/resourceGroups/trn12323resourcegroup/providers/Microsoft.Network/virtualNetworks/trn12323vnet/subnets/trn12323subnet", New:"", NewComputed:false, NewRemoved:true, NewExtra:interface {}(nil), RequiresNew:false, Type:0x0}, "ip_configuration.~463970659.private_ip_address":*terraform.ResourceAttrDiff{Old:"", New:"", NewComputed:true, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Type:0x0}, "ip_configuration.~463970659.load_balancer_inbound_nat_rules_ids.#":*terraform.ResourceAttrDiff{Old:"", New:"", NewComputed:true, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Type:0x0}, "ip_configuration.~463970659.public_ip_address_id":*terraform.ResourceAttrDiff{Old:"", New:"${element(azurerm_public_ip.publicip.*.id, count.index)}", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Type:0x0}, "ip_configuration.~463970659.name":*terraform.ResourceAttrDiff{Old:"", New:"testconfiguration1", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Type:0x0}, "ip_configuration.463970659.name":*terraform.ResourceAttrDiff{Old:"testconfiguration1", New:"", NewComputed:false, NewRemoved:true, NewExtra:interface {}(nil), RequiresNew:false, Type:0x0}}, Destroy:false, DestroyTainted:false}
    Diff Two (usually from apply): *terraform.InstanceDiff{Attributes:map[string]*terraform.ResourceAttrDiff(nil), Destroy:false, DestroyTainted:false}

Also include as much context as you can about your config, state, and the steps you performed to trigger this error.

Terraform does not automatically rollback in the face of errors.
Instead, your Terraform state file has been partially updated with
any resources that successfully completed. Please address the error
above and apply again to incrementally change your infrastructure.

### Expected Behavior
Changing the count variable adds more network interfaces

### Actual Behavior
Got above error

### Steps to Reproduce
Please list the steps required to reproduce the issue, for example:
1. `terraform apply`
2. change template variable "count" { default = "1" } => variable "count" { default = "2" }
3. `terraform apply`

**Comment by stack72** [comment]
Hi @pgroene 

We believe this is fixed in 0.7 since we made a lot of changes to how the ordering of lists are created. The proof is, of course, in the example:

```
% terraform apply
azurerm_resource_group.trainingvms: Creating...
  location: "" => "westeurope"
  name:     "" => "trn12323resourcegroup"
azurerm_resource_group.trainingvms: Creation complete
azurerm_virtual_network.vnet: Creating...
  address_space.#:     "" => "1"
  address_space.0:     "" => "10.0.0.0/16"
  location:            "" => "westeurope"
  name:                "" => "trn12323vnet"
  resource_group_name: "" => "trn12323resourcegroup"
  subnet.#:            "" => "<computed>"
azurerm_public_ip.publicip: Creating...
  fqdn:                         "" => "<computed>"
  ip_address:                   "" => "<computed>"
  location:                     "" => "westeurope"
  name:                         "" => "trn123230"
  public_ip_address_allocation: "" => "static"
  resource_group_name:          "" => "trn12323resourcegroup"
azurerm_public_ip.publicip: Still creating... (10s elapsed)
azurerm_virtual_network.vnet: Still creating... (10s elapsed)
azurerm_virtual_network.vnet: Still creating... (20s elapsed)
azurerm_public_ip.publicip: Still creating... (20s elapsed)
azurerm_public_ip.publicip: Creation complete
azurerm_virtual_network.vnet: Creation complete
azurerm_subnet.subnet: Creating...
  address_prefix:            "" => "10.0.2.0/24"
  ip_configurations.#:       "" => "<computed>"
  name:                      "" => "trn12323subnet"
  network_security_group_id: "" => "<computed>"
  resource_group_name:       "" => "trn12323resourcegroup"
  route_table_id:            "" => "<computed>"
  virtual_network_name:      "" => "trn12323vnet"
azurerm_subnet.subnet: Still creating... (10s elapsed)
azurerm_subnet.subnet: Still creating... (20s elapsed)
azurerm_subnet.subnet: Creation complete
azurerm_network_interface.nic: Creating...
  applied_dns_servers.#:                                                 "" => "<computed>"
  dns_servers.#:                                                         "" => "<computed>"
  enable_ip_forwarding:                                                  "" => "false"
  internal_dns_name_label:                                               "" => "<computed>"
  internal_fqdn:                                                         "" => "<computed>"
  ip_configuration.#:                                                    "" => "1"
  ip_configuration.1837620067.load_balancer_backend_address_pools_ids.#: "" => "<computed>"
  ip_configuration.1837620067.load_balancer_inbound_nat_rules_ids.#:     "" => "<computed>"
  ip_configuration.1837620067.name:                                      "" => "testconfiguration1"
  ip_configuration.1837620067.private_ip_address:                        "" => "<computed>"
  ip_configuration.1837620067.private_ip_address_allocation:             "" => "dynamic"
  ip_configuration.1837620067.public_ip_address_id:                      "" => "/subscriptions/34ca515c-4629-458e-bf7c-738d77e0d0ea/resourceGroups/trn12323resourcegroup/providers/Microsoft.Network/publicIPAddresses/trn123230"
  ip_configuration.1837620067.subnet_id:                                 "" => "/subscriptions/34ca515c-4629-458e-bf7c-738d77e0d0ea/resourceGroups/trn12323resourcegroup/providers/Microsoft.Network/virtualNetworks/trn12323vnet/subnets/trn12323subnet"
  location:                                                              "" => "westeurope"
  mac_address:                                                           "" => "<computed>"
  name:                                                                  "" => "trn12323nic0"
  network_security_group_id:                                             "" => "<computed>"
  private_ip_address:                                                    "" => "<computed>"
  resource_group_name:                                                   "" => "trn12323resourcegroup"
  virtual_machine_id:                                                    "" => "<computed>"
azurerm_network_interface.nic: Creation complete

Apply complete! Resources: 5 added, 0 changed, 0 destroyed.
```

Then I made the following change:

```
 variable "count" { default = "1" } => variable "count" { default = "2" }
```

The new Terraform apply gave me the following:

```
% terraform apply
azurerm_resource_group.trainingvms: Refreshing state... (ID: /subscriptions/34ca515c-4629-458e-bf7c-738d77e0d0ea/resourceGroups/trn12323resourcegroup)
azurerm_virtual_network.vnet: Refreshing state... (ID: /subscriptions/34ca515c-4629-458e-bf7c-738d77e0d0ea/resourceGroups/trn12323resourcegroup/providers/Microsoft.Network/virtualNetworks/trn12323vnet)
azurerm_public_ip.publicip.0: Refreshing state... (ID: /subscriptions/34ca515c-4629-458e-bf7c-738d77e0d0ea/resourceGroups/trn12323resourcegroup/providers/Microsoft.Network/publicIPAddresses/trn123230)
azurerm_subnet.subnet: Refreshing state... (ID: /subscriptions/34ca515c-4629-458e-bf7c-738d77e0d0ea/resourceGroups/trn12323resourcegroup/providers/Microsoft.Network/virtualNetworks/trn12323vnet/subnets/trn12323subnet)
azurerm_network_interface.nic.0: Refreshing state... (ID: /subscriptions/34ca515c-4629-458e-bf7c-738d77e0d0ea/resourceGroups/trn12323resourcegroup/providers/Microsoft.Network/networkInterfaces/trn12323nic0)
azurerm_public_ip.publicip.1: Creating...
  fqdn:                         "" => "<computed>"
  ip_address:                   "" => "<computed>"
  location:                     "" => "westeurope"
  name:                         "" => "trn123231"
  public_ip_address_allocation: "" => "static"
  resource_group_name:          "" => "trn12323resourcegroup"
azurerm_public_ip.publicip.1: Still creating... (10s elapsed)
azurerm_public_ip.publicip.1: Still creating... (20s elapsed)
azurerm_public_ip.publicip.1: Creation complete
azurerm_network_interface.nic.1: Creating...
  applied_dns_servers.#:                                                 "" => "<computed>"
  dns_servers.#:                                                         "" => "<computed>"
  enable_ip_forwarding:                                                  "" => "false"
  internal_dns_name_label:                                               "" => "<computed>"
  internal_fqdn:                                                         "" => "<computed>"
  ip_configuration.#:                                                    "" => "1"
  ip_configuration.1956440610.load_balancer_backend_address_pools_ids.#: "" => "<computed>"
  ip_configuration.1956440610.load_balancer_inbound_nat_rules_ids.#:     "" => "<computed>"
  ip_configuration.1956440610.name:                                      "" => "testconfiguration1"
  ip_configuration.1956440610.private_ip_address:                        "" => "<computed>"
  ip_configuration.1956440610.private_ip_address_allocation:             "" => "dynamic"
  ip_configuration.1956440610.public_ip_address_id:                      "" => "/subscriptions/34ca515c-4629-458e-bf7c-738d77e0d0ea/resourceGroups/trn12323resourcegroup/providers/Microsoft.Network/publicIPAddresses/trn123231"
  ip_configuration.1956440610.subnet_id:                                 "" => "/subscriptions/34ca515c-4629-458e-bf7c-738d77e0d0ea/resourceGroups/trn12323resourcegroup/providers/Microsoft.Network/virtualNetworks/trn12323vnet/subnets/trn12323subnet"
  location:                                                              "" => "westeurope"
  mac_address:                                                           "" => "<computed>"
  name:                                                                  "" => "trn12323nic1"
  network_security_group_id:                                             "" => "<computed>"
  private_ip_address:                                                    "" => "<computed>"
  resource_group_name:                                                   "" => "trn12323resourcegroup"
  virtual_machine_id:                                                    "" => "<computed>"
azurerm_network_interface.nic.1: Creation complete

Apply complete! Resources: 2 added, 0 changed, 0 destroyed.
```

As you can see, it only added the 2 new resources, IP Address and NIC

0.7 should be out soon! Hope this is ok

Thanks

Paul

**Comment by pgroene** [comment]
thx that is good enough for me!
